core:ppc clang/xlC build failure

- OpenCV => master
- Compiler => clang tested on (4.0.1-6, 5.0.0-2, 6.0.0-svn310776-1),
xlc, tested on (13.1.5.1)

on clang:
> /home/builder/opencv/modules/core/include/opencv2/core/vsx_utils.hpp:667:1: error: call to 'vec_promote' is ambiguous
VSX_IMPL_LOAD_L8(vec_double2, double)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/builder/opencv/modules/core/include/opencv2/core/vsx_utils.hpp:651:17: note: expanded from macro 'VSX_IMPL_LOAD_L8'
{ return ((Tvec)vec_promote(*((uint64*)p), 0)); }                               \
                ^~~~~~~~~~~
/usr/include/clang/4.0.1/include/altivec.h:13612:51: note: candidate function
static __inline__ vector signed char __ATTRS_o_ai vec_promote(signed char __a,
                                                  ^
/usr/include/clang/4.0.1/include/altivec.h:13620:1: note: candidate function
vec_promote(unsigned char __a, int __b) {
^
/usr/include/clang/4.0.1/include/altivec.h:13626:45: note: candidate function
static __inline__ vector short __ATTRS_o_ai vec_promote(short __a, int __b) {
                                            ^
/usr/include/clang/4.0.1/include/altivec.h:13633:1: note: candidate function
vec_promote(unsigned short __a, int __b) {
^
/usr/include/clang/4.0.1/include/altivec.h:13639:43: note: candidate function
static __inline__ vector int __ATTRS_o_ai vec_promote(int __a, int __b) {
                                          ^
/usr/include/clang/4.0.1/include/altivec.h:13645:52: note: candidate function
static __inline__ vector unsigned int __ATTRS_o_ai vec_promote(unsigned int __a,
                                                   ^
/usr/include/clang/4.0.1/include/altivec.h:13652:45: note: candidate function
static __inline__ vector float __ATTRS_o_ai vec_promote(float __a, int __b) {}

-----------------------

on xlc: 

> /home/builder/opencv/modules/core/include/opencv2/core/vsx_utils.hpp:666:1: error: call to 'vec_promote' is ambiguous
VSX_IMPL_LOAD_L8(vec_dword2,  int64)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/builder/opencv/modules/core/include/opencv2/core/vsx_utils.hpp:651:17: note: expanded from macro 'VSX_IMPL_LOAD_L8'
{ return ((Tvec)vec_promote(*((uint64*)p), 0)); }                               \
                ^~~~~~~~~~~
/opt/ibm/xlC/13.1.5/include/altivec.h:1439:1: note: candidate function
vec_promote(double __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_f8_Vddi");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1442:1: note: candidate function
vec_promote(signed long long __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_s8_VSlli");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1445:1: note: candidate function
vec_promote(unsigned long long __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_u8_VUlli");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1448:1: note: candidate function
vec_promote(signed int __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_s4_VSiii");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1451:1: note: candidate function
vec_promote(unsigned int __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_u4_VUiii");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1454:1: note: candidate function
vec_promote(signed short __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_s2_VSssi");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1457:1: note: candidate function
vec_promote(unsigned short __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_u2_VUssi");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1460:1: note: candidate function
vec_promote(signed char __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_s1_VScci");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:1463:1: note: candidate function
vec_promote(unsigned char __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_u1_VUcci");
^
/opt/ibm/xlC/13.1.5/include/altivec.h:6298:1: note: candidate function
vec_promote(float __a, signed int __b) __asm__("__IBMBuiltin_vec_promote_f4_Vffi");

- OpenCV => master
- Compiler => xlc tested on (13.1.5.1)

> /home/builder/opencv/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:641:19: error: 1540-2882 The built-in function
      "__IBMBuiltin_vec_sld_VUcVUcVUci"'s argument "3" must be in the range "0...15".
    return _Tpvec(vec_sld(b.val, a.val, CV_SHIFT));

- OpenCV => All
- Compiler => xlC tested on (13.1.5.1)

> /home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20099:11: error: comparison between pointer and integer ('int' and 'int *')
  if (lhs == rhs) {
      ~~~ ^  ~~~
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20128:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<int, int *>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/home/builder/opencv/modules/core/test/test_ptr.cpp:81:5: note: in instantiation of function template specialization
      'testing::internal::EqHelper<false>::Compare<int, int *>' requested here
    EXPECT_EQ(NULL, p.get());
    ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20988:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20099:11: error: comparison between pointer and integer ('int' and 'void *')
  if (lhs == rhs) {
      ~~~ ^  ~~~
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20128:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<int, void *>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/home/builder/opencv/modules/core/test/test_ptr.cpp:105:9: note: in instantiation of function template specialization
      'testing::internal::EqHelper<false>::Compare<int, void *>' requested here
        EXPECT_EQ(NULL, p.get());
        ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20988:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20099:11: error: comparison between pointer and integer ('int' and
      '(anonymous namespace)::Reporter *')
  if (lhs == rhs) {
      ~~~ ^  ~~~
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20128:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<int, (anonymous namespace)::Reporter *>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/home/builder/opencv/modules/core/test/test_ptr.cpp:190:5: note: in instantiation of function template specialization
      'testing::internal::EqHelper<false>::Compare<int, (anonymous namespace)::Reporter *>' requested here
    EXPECT_EQ(NULL, p1.get());
    ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20988:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20099:11: error: comparison between pointer and integer ('int' and
      '(anonymous namespace)::SubReporter *')
  if (lhs == rhs) {
      ~~~ ^  ~~~
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20128:12: note: in instantiation of function template specialization
      'testing::internal::CmpHelperEQ<int, (anonymous namespace)::SubReporter *>' requested here
    return CmpHelperEQ(lhs_expression, rhs_expression, lhs, rhs);
           ^
/home/builder/opencv/modules/core/test/test_ptr.cpp:330:9: note: in instantiation of function template specialization
      'testing::internal::EqHelper<false>::Compare<int, (anonymous namespace)::SubReporter *>' requested here
        EXPECT_EQ(NULL, p2.get());
        ^
/home/builder/opencv/modules/ts/include/opencv2/ts/ts_gtest.h:20988:63: note: expanded from macro 'EXPECT_EQ'
                      EqHelper<GTEST_IS_NULL_LITERAL_(val1)>::Compare, \
                                                              ^
4 errors generated.
Error while processing /home/builder/opencv/modules/core/test/test_ptr.cpp.

--- a/modules/core/include/opencv2/core/hal/intrin_vsx.hpp
+++ b/modules/core/include/opencv2/core/hal/intrin_vsx.hpp
@@ -638,7 +638,11 @@ inline _Tpvec v_rotate_right(const _Tpvec& a, const _Tpvec& b)
     enum { CV_SHIFT = 16 - imm * (sizeof(typename _Tpvec::lane_type)) };
     if (CV_SHIFT == 16)
         return a;
+#ifdef __IBMCPP__
+    return _Tpvec(vec_sld(b.val, a.val, CV_SHIFT & 15));
+#else
     return _Tpvec(vec_sld(b.val, a.val, CV_SHIFT));
+#endif
 }
 
 template<int imm, typename _Tpvec>

--- a/modules/core/include/opencv2/core/vsx_utils.hpp
+++ b/modules/core/include/opencv2/core/vsx_utils.hpp
@@ -420,6 +420,21 @@ FORCE_INLINE(rt) fnm(const rg& a) { return __builtin_convertvector(a, rt); }
     { return vec_div(vec_double2_sp(1), vec_sqrt(a)); }
 #endif
 
+// vec_promote missing support for doubleword
+FORCE_INLINE(vec_dword2) vec_promote(long long a, int b)
+{
+    vec_dword2 ret = vec_dword2_z;
+    ret[b & 1] = a;
+    return ret;
+}
+
+FORCE_INLINE(vec_udword2) vec_promote(unsigned long long a, int b)
+{
+    vec_udword2 ret = vec_udword2_z;
+    ret[b & 1] = a;
+    return ret;
+}
+
 // vec_popcnt should return unsigned but clang has different thought just like gcc in vec_vpopcnt
 #define VSX_IMPL_POPCNTU(Tvec, Tvec2, ucast)   \
 FORCE_INLINE(Tvec) vec_popcntu(const Tvec2& a) \
@@ -569,6 +584,12 @@ VSX_IMPL_DIRTY_ODD(vec_udword2, vec_float4, vec_ctulo, vec_ctul)
 
     FORCE_INLINE(vec_dword2) vec_splats(int64 v)
     { return vec_splats((long long) v); }
+
+    FORCE_INLINE(vec_udword2) vec_promote(uint64 a, int b)
+    { return vec_promote((unsigned long long) a, b); }
+
+    FORCE_INLINE(vec_dword2) vec_promote(int64 a, int b)
+    { return vec_promote((long long) a, b); }
 #endif
 
 /*

--- a/modules/core/test/test_ptr.cpp
+++ b/modules/core/test/test_ptr.cpp
@@ -41,6 +41,12 @@
 
 #include "test_precomp.hpp"
 
+#ifdef GTEST_CAN_COMPARE_NULL
+#  define EXPECT_NULL(ptr) EXPECT_EQ(NULL, ptr)
+#else
+#  define EXPECT_NULL(ptr) EXPECT_TRUE(ptr == NULL)
+#endif
+
 using namespace cv;
 
 namespace {
@@ -78,7 +84,7 @@ int dummyObject;
 TEST(Core_Ptr, default_ctor)
 {
     Ptr<int> p;
-    EXPECT_EQ(NULL, p.get());
+    EXPECT_NULL(p.get());
 }
 
 TEST(Core_Ptr, owning_ctor)
@@ -102,7 +108,7 @@ TEST(Core_Ptr, owning_ctor)
 
     {
         Ptr<void> p((void*)0, ReportingDeleter(&deleted));
-        EXPECT_EQ(NULL, p.get());
+        EXPECT_NULL(p.get());
     }
 
     EXPECT_FALSE(deleted);
@@ -187,7 +193,7 @@ TEST(Core_Ptr, release)
     Ptr<Reporter> p1(new Reporter(&deleted));
     p1.release();
     EXPECT_TRUE(deleted);
-    EXPECT_EQ(NULL, p1.get());
+    EXPECT_NULL(p1.get());
 }
 
 TEST(Core_Ptr, reset)
@@ -253,7 +259,7 @@ TEST(Core_Ptr, accessors)
 {
     {
         Ptr<int> p;
-        EXPECT_EQ(NULL, static_cast<int*>(p));
+        EXPECT_NULL(static_cast<int*>(p));
         EXPECT_TRUE(p.empty());
     }
 
@@ -327,7 +333,7 @@ TEST(Core_Ptr, casts)
     {
         Ptr<Reporter> p1(new Reporter(&deleted));
         Ptr<SubReporter> p2 = p1.dynamicCast<SubReporter>();
-        EXPECT_EQ(NULL, p2.get());
+        EXPECT_NULL(p2.get());
         p1.release();
         EXPECT_FALSE(deleted);
     }

